### DISCUSSION TITLE
ActiveJob & Pro Features: A rock and a hard place

### DISCUSSION THREAD

**Discussion by bmulholland** [sidekiq/sidekiq#5100] [General]
(Writing this as someone who just paid for a month of Sidekiq Pro)

I'm pulled toward the advanced/pro features, and want the best reliability that we can get. As I understand it, that means [using `Sidekiq::Worker` instead of ActiveJob](https://github.com/mperham/sidekiq/wiki/Active-Job#commercial-features).

Unfortunately, migrating to Sidekiq Workers is a significant ergonomic/developer experience loss, especially considering we use Sorbet. Since we'd be losing serialization/global IDs _and_ keyword args, we'd go from:

```ruby
sig { params(
  activerecord_obj1: ActiveRecordModel1,
  activerecord_obj2: T.nilable(ActiveRecordModel2),
  optional_flag: T::Boolean
).void }
def perform(activerecord_obj1, activerecord_obj2: nil, optional_flag: false)
```

That we can call with clarity: `Job.perform_later(obj1, activerecord_obj2: obj2, optional_flag: true)`
(This is far more readable for actual examples, where the keyword provides important context to the reader)

To:
```ruby
sig { params(
  activerecord_obj1_id: Integer,
  activerecord_obj2_id: T.nilable(Integer),
  optional_flag: T::Boolean
).void }
def perform(activerecord_obj1_id, activerecord_obj2_id = null, optional_flag = false)
activerecord_obj1 = ActiveRecordModel1.find(activerecord_obj1_id)
activerecord_obj2 = ActiveRecordModel2.find(activerecord_obj2_id) if activerecord_obj2_id.present?
```

That:
1. Is a much-less-readable (because it has far less context) `Worker.perform_async(obj1.id, obj2.id, true)`.
2. Loses typechecking on the object type.
3. Is yet more boilerplate I hate writing, especially when we know a computer can just do it for me.
4. Likely requires individual error handling that is easy to be written differently in each worker.

I've already read the threads on this, and I know the intent is to keep the Sidekiq API JSON encodable. That's a fine goal, and I get it, but that doesn't change the reality above. This especially applies because, as a ruby-only user, I don't care about interoperability. I'd even be fine with an optional extension that sits on top of sidekiq to provide these things, but as far as I can tell [they don't exist](https://github.com/mperham/sidekiq/wiki/Related-Projects).

As it is, I feel stuck. Without these issues I'd surely go all in on Sidekiq Workers. With them, I'm being forced to weigh a risk of bugs, lack of clarity (which I value above almost everything else), and extra unneeded code versus improved stability and advanced features that may (or may not) outweigh the tradeoffs I've outlined. Overall, it strikes me as backwards that adopting the paid/Pro features requires worse DX.

So:

Are there any plans to improve support for ActiveJob and [change your recommendation about not using it with Batches](https://github.com/mperham/sidekiq/issues/4637#issuecomment-656199172) etc?
Alternatively, are there any plans to add support for Global IDs and [change your recommendation to use ActiveJob for that](https://github.com/mperham/sidekiq/wiki/Related-Projects)?
And what about keyword args -- could [one of the workarounds](https://github.com/mperham/sidekiq/issues/2372#issuecomment-998092507) be officially adopted?

**Comment by mperham** [comment]
Just to be clear, all of Sidekiq Pro’s reliability features work perfectly
fine with ActiveJob.

**Reply by bmulholland** [reply]
Okay, but what about batches? Other features?

**Comment by mperham** [comment]
You’re the first person I’ve ever heard from outside of Stripe or Shopify who is actually using Sorbet with their app. It’s not on my radar because there has been zero demand.

**Reply by bmulholland** [reply]
Yep, totally understand why it hasn't been considered before :)

**Comment by mperham** [comment]
AJ’s built-in error handling via rescue_from breaks batches. If you use
simple AJs with Sidekiq’s standard retry support, batches  should work fine
AFAIK.

**Reply by bmulholland** [reply]
Thanks, that's helpful.

Any thoughts on the questions I posted at the end?

**Reply by mperham** [reply]
There are numerous social and technical reasons why I have not provided any more complex argument handling.

1. I’ve spent years training people to understand that sidekiq arguments equal json.
2. Pure Json is the most widely compatible format and does not require any rails gem dependency.
2. Complex argument handling can raise errors, and so now you require a more complex error handling system. This is the reason why Rails had to create rescue_from in the first place, because a global ID look up can raise a database exception.
3. There are third-party sidekiq gems which add keyboard argument and global ID support. You are welcome to try and use one of those.

After thinking more about it I might be able to provide global ID support in a reasonable way. I need to do more research.

**Reply by bmulholland** [reply]
Thanks for explaining -- I totally understand why you'd want to keep it simple :).

Re: the third-party gems, do you know any of those offhand? I didn't see any in the related projects list.
For Global ID I found https://github.com/bolshakov/sidekiq-global_id, though it's archived so I'm hesitant.
For keyword args I've only found the workarounds I linked in my original post.

Thanks again for your thoughts and happy holidays :)

**Comment by mperham** [comment]
Yeah, that’s the gem I was thinking of. If you look at the impl, it just bolts on AJ arg ser/deserialization. I’m not sure that’s a great idea but some people use that gem and I’ve not heard of any problems with it.

I don’t have any suggestions for kwargs; I never use them personally.

**Reply by bmulholland** [reply]
Thanks, that's all good to know. Appreciate your thoughts, hope you're having a nice holiday :)
